File: app/src/test/java/com/slapshotapps/dragonshockey/activities/careerStats/CareerStatsVMTest.java
package com.slapshotapps.dragonshockey.activities.careerStats;

import com.slapshotapps.dragonshockey.models.GameStats;
import com.slapshotapps.dragonshockey.models.Player;
import com.slapshotapps.dragonshockey.models.PlayerPosition;
import com.slapshotapps.dragonshockey.models.PlayerStats;
import com.slapshotapps.dragonshockey.models.SeasonStats;
import java.util.ArrayList;
import java.util.List;
import org.junit.Test;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.Is.is;

public class CareerStatsVMTest {

    @Test
    public void testPlayerName() {

        Player player = new Player();
        player.setFirstName("Bob");
        player.setLastName("Builder");
        player.setPlayerID(9);

        CareerStatsVM careerStatsVM = new CareerStatsVM(player, null, null);

        assertThat(careerStatsVM.getPlayerName(), is("Bob Builder"));
    }

    @Test
    public void testPlayerNameNull() {

        Player player = new Player();
        player.setPlayerID(9);

        CareerStatsVM careerStatsVM = new CareerStatsVM(player, null, null);

        assertThat(careerStatsVM.getPlayerName(), is(""));
    }

    @Test
    public void testPlayerNumber() {
        Player player = new Player();
        player.setFirstName("Bob");
        player.setLastName("Builder");
        player.setPlayerID(9);
        player.setNumber(99);

        CareerStatsVM careerStatsVM = new CareerStatsVM(player, null, null);

        assertThat(careerStatsVM.getPlayerNumber(), is("# 99"));
    }

    @Test
    public void testPlayerNumberEmpty() {
        Player player = new Player();
        player.setFirstName("Bob");
        player.setLastName("Builder");
        player.setPlayerID(9);

        CareerStatsVM careerStatsVM = new CareerStatsVM(player, null, null);

        assertThat(careerStatsVM.getPlayerNumber(), is("# 0"));
    }

    @Test
    public void testPlayerPosition() {
        Player player = new Player("F");
        player.setFirstName("Bob");
        player.setLastName("Builder");
        player.setPlayerID(9);
        player.setNumber(99);

        CareerStatsVM careerStatsVM = new CareerStatsVM(player, null, null);

        assertThat(careerStatsVM.getPlayerPosition(), is("Forward"));
    }

    @Test
    public void testNonGoalieCurrentSeasonAdding() {

        PlayerStats playerStats = new PlayerStats(2, "bob", "Builder", PlayerPosition.FORWARD);
        playerStats.setGoals(2);
        playerStats.setAssists(4);
        playerStats.setPenaltyMinutes(5);
        playerStats.setGamesPlayed(10);

        Player player = new Player("F");
        player.setFirstName("Bob");
        player.setLastName("Builder");
        player.setPlayerID(2);
        player.setNumber(99);

        CareerStatsVM careerStatsVM = new CareerStatsVM(player, playerStats, null);

        List<PlayerSeasonStatsVM> playerSeasonStats = careerStatsVM.getStats();

        assertThat(playerSeasonStats.size(), is(2));
        assertThat(playerSeasonStats.get(0).getGoals(), is(2));
        assertThat(playerSeasonStats.get(0).getAssists(), is(4));
        assertThat(playerSeasonStats.get(0).getPenaltyMinutes(), is(5));
        assertThat(playerSeasonStats.get(0).getPoints(), is(String.valueOf(6)));

        //career
        assertThat(playerSeasonStats.get(1).getGoals(), is(2));
        assertThat(playerSeasonStats.get(1).getAssists(), is(4));
        assertThat(playerSeasonStats.get(1).getPenaltyMinutes(), is(5));
        assertThat(playerSeasonStats.get(1).getPoints(), is(String.valueOf(6)));
        assertThat(playerSeasonStats.get(1).getSeasonID(), is("Career"));
    }

    @Test
    public void testGoalieCurrentSeasonAdding() {

        PlayerStats playerStats = new PlayerStats(2, "bob", "Builder", PlayerPosition.GOALIE);
        playerStats.setPenaltyMinutes(5);
        playerStats.setGamesPlayed(10);
        playerStats.setGoalsAgainst(18);

        Player player = new Player("G");
        player.setFirstName("Bob");
        player.setLastName("Builder");
        player.setPlayerID(2);
        player.setNumber(99);

        CareerStatsVM careerStatsVM = new CareerStatsVM(player, playerStats, null);

        List<PlayerSeasonStatsVM> playerSeasonStats = careerStatsVM.getStats();

        assertThat(playerSeasonStats.size(), is(2));
        assertThat(playerSeasonStats.get(0).getPenaltyMinutes(), is(5));
        assertThat(playerSeasonStats.get(0).getGoalsAgainst(), is(String.valueOf(18)));
        assertThat(playerSeasonStats.get(0).getGoalsAgainstAverage(), is("1.80"));
        assertThat(playerSeasonStats.get(0).getGamesPlayed(), is(String.valueOf(10)));

        //career
        assertThat(playerSeasonStats.get(1).getGamesPlayed(), is(String.valueOf(10)));
        assertThat(playerSeasonStats.get(1).getGoalsAgainst(), is(String.valueOf(18)));
        assertThat(playerSeasonStats.get(1).getGoalsAgainstAverage(), is("1.80"));
        assertThat(playerSeasonStats.get(1).getPenaltyMinutes(), is(5));
        assertThat(playerSeasonStats.get(1).getSeasonID(), is("Career"));
    }

    @Test
    public void testNonGoalieHistoricalSeasonFiltering() {

        List<SeasonStats> seasonStats = new ArrayList<>();
        seasonStats.add(new SeasonStats());
        seasonStats.add(new SeasonStats());

        for (int i = 0; i < 5; i++) {
            seasonStats.get(0).getStats().add(new GameStats());
            seasonStats.get(0).getStats().get(i).setGameStats(getTeamStatsForGame());

            seasonStats.get(1).getStats().add(new GameStats());
            seasonStats.get(1).getStats().get(i).setGameStats(getTeamStatsForGame());
        }

        Player player = new Player("F");
        player.setFirstName("Bob");
        player.setLastName("Builder");
        player.setPlayerID(2);
        player.setNumber(99);

        CareerStatsVM careerStatsVM = new CareerStatsVM(player, null, seasonStats);

        List<PlayerSeasonStatsVM> playerSeasonStats = careerStatsVM.getStats();

        assertThat(playerSeasonStats.size(), is(3));
        assertThat(playerSeasonStats.get(0).getGoals(), is(15));
        assertThat(playerSeasonStats.get(0).getAssists(), is(20));
        assertThat(playerSeasonStats.get(0).getPenaltyMinutes(), is(10));
        assertThat(playerSeasonStats.get(0).getPoints(), is(String.valueOf(35)));

        assertThat(playerSeasonStats.get(1).getGoals(), is(15));
        assertThat(playerSeasonStats.get(1).getAssists(), is(20));
        assertThat(playerSeasonStats.get(1).getPenaltyMinutes(), is(10));
        assertThat(playerSeasonStats.get(1).getPoints(), is(String.valueOf(35)));

        assertThat(playerSeasonStats.get(2).getGoals(), is(30));
        assertThat(playerSeasonStats.get(2).getAssists(), is(40));
        assertThat(playerSeasonStats.get(2).getPenaltyMinutes(), is(20));
        assertThat(playerSeasonStats.get(2).getPoints(), is(String.valueOf(70)));
    }

    @Test
    public void testGoalieHistoricalSeasonFiltering() {

        List<SeasonStats> seasonStats = new ArrayList<>();
        seasonStats.add(new SeasonStats());
        seasonStats.add(new SeasonStats());

        for (int i = 0; i < 5; i++) {
            seasonStats.get(0).getStats().add(new GameStats());
            seasonStats.get(0).getStats().get(i).setGameStats(getTeamStatsForGame());

            seasonStats.get(1).getStats().add(new GameStats());
            seasonStats.get(1).getStats().get(i).setGameStats(getTeamStatsForGame());
        }

        Player player = new Player("G");
        player.setFirstName("Bob");
        player.setLastName("Builder");
        player.setPlayerID(4);
        player.setNumber(99);

        CareerStatsVM careerStatsVM = new CareerStatsVM(player, null, seasonStats);

        List<PlayerSeasonStatsVM> playerSeasonStats = careerStatsVM.getStats();

        assertThat(playerSeasonStats.size(), is(3));
        assertThat(playerSeasonStats.get(0).getPenaltyMinutes(), is(10));
        assertThat(playerSeasonStats.get(0).getGoalsAgainst(), is("20"));
        assertThat(playerSeasonStats.get(0).getGoalsAgainstAverage(), is("4.00"));
        assertThat(playerSeasonStats.get(0).getGamesPlayed(), is("5"));



        assertThat(playerSeasonStats.get(1).getPenaltyMinutes(), is(10));
        assertThat(playerSeasonStats.get(1).getGoalsAgainst(), is("20"));
        assertThat(playerSeasonStats.get(1).getGoalsAgainstAverage(), is("4.00"));
        assertThat(playerSeasonStats.get(1).getGamesPlayed(), is("5"));

        assertThat(playerSeasonStats.get(2).getPenaltyMinutes(), is(20));
        assertThat(playerSeasonStats.get(2).getGoalsAgainst(), is("40"));
        assertThat(playerSeasonStats.get(2).getGoalsAgainstAverage(), is("4.00"));
        assertThat(playerSeasonStats.get(2).getGamesPlayed(), is("10"));

    }

    @Test
    public void testFullCareer() {
        List<SeasonStats> seasonStats = new ArrayList<>();
        seasonStats.add(new SeasonStats("Winter 16"));
        seasonStats.add(new SeasonStats("Fall 17"));

        for (int i = 0; i < 5; i++) {
            seasonStats.get(0).getStats().add(new GameStats());
            seasonStats.get(0).getStats().get(i).setGameStats(getTeamStatsForGame());

            seasonStats.get(1).getStats().add(new GameStats());
            seasonStats.get(1).getStats().get(i).setGameStats(getTeamStatsForGame());
        }

        Player player = new Player("F");
        player.setFirstName("Bob");
        player.setLastName("Builder");
        player.setPlayerID(2);
        player.setNumber(99);

        PlayerStats playerStats = new PlayerStats(2, "bob", "Builder", PlayerPosition.FORWARD);
        playerStats.setGoals(2);
        playerStats.setAssists(4);
        playerStats.setGamesPlayed(10);
        playerStats.setPenaltyMinutes(8);

        CareerStatsVM careerStatsVM = new CareerStatsVM(player, playerStats, seasonStats);

        List<PlayerSeasonStatsVM> playerSeasonStats = careerStatsVM.getStats();

        assertThat(playerSeasonStats.size(), is(4));

        assertThat(playerSeasonStats.get(0).getGoals(), is(15));
        assertThat(playerSeasonStats.get(0).getAssists(), is(20));
        assertThat(playerSeasonStats.get(0).getPenaltyMinutes(), is(10));
        assertThat(playerSeasonStats.get(0).getPoints(), is(String.valueOf(35)));
        assertThat(playerSeasonStats.get(0).getSeasonID(), is("Winter 16"));

        assertThat(playerSeasonStats.get(1).getGoals(), is(15));
        assertThat(playerSeasonStats.get(1).getAssists(), is(20));
        assertThat(playerSeasonStats.get(1).getPenaltyMinutes(), is(10));
        assertThat(playerSeasonStats.get(1).getPoints(), is(String.valueOf(35)));
        assertThat(playerSeasonStats.get(1).getSeasonID(), is("Fall 17"));

        assertThat(playerSeasonStats.get(2).getGoals(), is(2));
        assertThat(playerSeasonStats.get(2).getAssists(), is(4));
        assertThat(playerSeasonStats.get(2).getPenaltyMinutes(), is(8));
        assertThat(playerSeasonStats.get(2).getPoints(), is(String.valueOf(6)));
        assertThat(playerSeasonStats.get(2).getSeasonID(), is("Current"));

        assertThat(playerSeasonStats.get(3).getGoals(), is(32));
        assertThat(playerSeasonStats.get(3).getAssists(), is(44));
        assertThat(playerSeasonStats.get(3).getPenaltyMinutes(), is(28));
        assertThat(playerSeasonStats.get(3).getPoints(), is(String.valueOf(76)));
        assertThat(playerSeasonStats.get(3).getSeasonID(), is("Career"));
    }

    private List<GameStats.Stats> getTeamStatsForGame() {

        List<GameStats.Stats> stats = new ArrayList<GameStats.Stats>();

        int goals = 1;
        int assists = 2;
        int goalsAgainst = 0;
        for (int playerID = 0; playerID < 10; playerID++) {
            stats.add(new GameStats.Stats(playerID, assists++, goals++, 2, true, goalsAgainst++));
        }

        return stats;
    }
}

File: app/src/test/java/com/slapshotapps/dragonshockey/activities/careerStats/CareerStatsVMTest.kt
package com.slapshotapps.dragonshockey.activities.careerStats

import com.slapshotapps.dragonshockey.models.*
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.core.Is.`is`
import org.junit.Test
import java.util.*

class CareerStatsVMTest {

    private val teamStatsForGame: List<GameStats.Stats>
        get() {

            val stats = ArrayList<GameStats.Stats>()

            var goals = 1
            var assists = 2
            var goalsAgainst = 0
            for (playerID in 0..9) {
                stats.add(GameStats.Stats(playerID, assists++, goals++, 2, true, goalsAgainst++))
            }

            return stats
        }

    @Test
    fun testPlayerName() {

        val player = Player()
        player.firstName = "Bob"
        player.lastName = "Builder"
        player.playerID = 9

        val careerStatsVM = CareerStatsVM(player, null, null)

        assertThat(careerStatsVM.playerName, `is`("Bob Builder"))
    }

    @Test
    fun testPlayerNameNull() {

        val player = Player()
        player.playerID = 9

        val careerStatsVM = CareerStatsVM(player, null, null)

        assertThat(careerStatsVM.playerName, `is`(""))
    }

    @Test
    fun testPlayerNumber() {
        val player = Player()
        player.firstName = "Bob"
        player.lastName = "Builder"
        player.playerID = 9
        player.number = 99

        val careerStatsVM = CareerStatsVM(player, null, null)

        assertThat(careerStatsVM.playerNumber, `is`("# 99"))
    }

    @Test
    fun testPlayerNumberEmpty() {
        val player = Player()
        player.firstName = "Bob"
        player.lastName = "Builder"
        player.playerID = 9

        val careerStatsVM = CareerStatsVM(player, null, null)

        assertThat(careerStatsVM.playerNumber, `is`("# 0"))
    }

    @Test
    fun testPlayerPosition() {
        val player = Player("F")
        player.firstName = "Bob"
        player.lastName = "Builder"
        player.playerID = 9
        player.number = 99

        val careerStatsVM = CareerStatsVM(player, null, null)

        assertThat(careerStatsVM.playerPosition, `is`("Forward"))
    }

    @Test
    fun testNonGoalieCurrentSeasonAdding() {

        val playerStats = PlayerStats(2, "bob", "Builder", PlayerPosition.FORWARD)
        playerStats.goals = 2
        playerStats.assists = 4
        playerStats.penaltyMinutes = 5
        playerStats.gamesPlayed = 10

        val player = Player("F")
        player.firstName = "Bob"
        player.lastName = "Builder"
        player.playerID = 2
        player.number = 99

        val careerStatsVM = CareerStatsVM(player, playerStats, null)

        val playerSeasonStats = careerStatsVM.stats

        assertThat(playerSeasonStats.size, `is`(2))
        assertThat(playerSeasonStats[0].goals, `is`(2))
        assertThat(playerSeasonStats[0].assists, `is`(4))
        assertThat(playerSeasonStats[0].penaltyMinutes, `is`(5))
        assertThat(playerSeasonStats[0].points, `is`(6.toString()))

        //career
        assertThat(playerSeasonStats[1].goals, `is`(2))
        assertThat(playerSeasonStats[1].assists, `is`(4))
        assertThat(playerSeasonStats[1].penaltyMinutes, `is`(5))
        assertThat(playerSeasonStats[1].points, `is`(6.toString()))
        assertThat(playerSeasonStats[1].seasonID, `is`("Career"))
    }

    @Test
    fun testGoalieCurrentSeasonAdding() {

        val playerStats = PlayerStats(2, "bob", "Builder", PlayerPosition.GOALIE)
        playerStats.penaltyMinutes = 5
        playerStats.gamesPlayed = 10
        playerStats.goalsAgainst = 18

        val player = Player("G")
        player.firstName = "Bob"
        player.lastName = "Builder"
        player.playerID = 2
        player.number = 99

        val careerStatsVM = CareerStatsVM(player, playerStats, null)

        val playerSeasonStats = careerStatsVM.stats

        assertThat(playerSeasonStats.size, `is`(2))
        assertThat(playerSeasonStats[0].penaltyMinutes, `is`(5))
        assertThat(playerSeasonStats[0].goalsAgainst, `is`(18))
        assertThat(playerSeasonStats[0].goalsAgainstAverage, `is`("1.80"))
        assertThat(playerSeasonStats[0].gamesPlayed, `is`(10))

        //career
        assertThat(playerSeasonStats[1].gamesPlayed, `is`(10))
        assertThat(playerSeasonStats[1].goalsAgainst, `is`(18))
        assertThat(playerSeasonStats[1].goalsAgainstAverage, `is`("1.80"))
        assertThat(playerSeasonStats[1].penaltyMinutes, `is`(5))
        assertThat(playerSeasonStats[1].seasonID, `is`("Career"))
    }

    @Test
    fun testNonGoalieHistoricalSeasonFiltering() {

        val seasonStats = ArrayList<SeasonStats>()
        seasonStats.add(SeasonStats())
        seasonStats.add(SeasonStats())

        for (i in 0..4) {
            seasonStats[0].stats.add(GameStats())
            seasonStats[0].stats[i].gameStats = teamStatsForGame

            seasonStats[1].stats.add(GameStats())
            seasonStats[1].stats[i].gameStats = teamStatsForGame
        }

        val player = Player("F")
        player.firstName = "Bob"
        player.lastName = "Builder"
        player.playerID = 2
        player.number = 99

        val careerStatsVM = CareerStatsVM(player, null, seasonStats)

        val playerSeasonStats = careerStatsVM.stats

        assertThat(playerSeasonStats.size, `is`(3))
        assertThat(playerSeasonStats[0].goals, `is`(15))
        assertThat(playerSeasonStats[0].assists, `is`(20))
        assertThat(playerSeasonStats[0].penaltyMinutes, `is`(10))
        assertThat(playerSeasonStats[0].points, `is`(35.toString()))

        assertThat(playerSeasonStats[1].goals, `is`(15))
        assertThat(playerSeasonStats[1].assists, `is`(20))
        assertThat(playerSeasonStats[1].penaltyMinutes, `is`(10))
        assertThat(playerSeasonStats[1].points, `is`(35.toString()))

        assertThat(playerSeasonStats[2].goals, `is`(30))
        assertThat(playerSeasonStats[2].assists, `is`(40))
        assertThat(playerSeasonStats[2].penaltyMinutes, `is`(20))
        assertThat(playerSeasonStats[2].points, `is`(70.toString()))
    }

    @Test
    fun testGoalieHistoricalSeasonFiltering() {

        val seasonStats = ArrayList<SeasonStats>()
        seasonStats.add(SeasonStats())
        seasonStats.add(SeasonStats())

        for (i in 0..4) {
            seasonStats[0].stats.add(GameStats())
            seasonStats[0].stats[i].gameStats = teamStatsForGame

            seasonStats[1].stats.add(GameStats())
            seasonStats[1].stats[i].gameStats = teamStatsForGame
        }

        val player = Player("G")
        player.firstName = "Bob"
        player.lastName = "Builder"
        player.playerID = 4
        player.number = 99

        val careerStatsVM = CareerStatsVM(player, null, seasonStats)

        val playerSeasonStats = careerStatsVM.stats

        assertThat(playerSeasonStats.size, `is`(3))
        assertThat(playerSeasonStats[0].penaltyMinutes, `is`(10))
        assertThat(playerSeasonStats[0].goalsAgainst, `is`(20))
        assertThat(playerSeasonStats[0].goalsAgainstAverage, `is`("4.00"))
        assertThat(playerSeasonStats[0].gamesPlayed, `is`(5))



        assertThat(playerSeasonStats[1].penaltyMinutes, `is`(10))
        assertThat(playerSeasonStats[1].goalsAgainst, `is`(20))
        assertThat(playerSeasonStats[1].goalsAgainstAverage, `is`("4.00"))
        assertThat(playerSeasonStats[1].gamesPlayed, `is`(5))

        assertThat(playerSeasonStats[2].penaltyMinutes, `is`(20))
        assertThat(playerSeasonStats[2].goalsAgainst, `is`(40))
        assertThat(playerSeasonStats[2].goalsAgainstAverage, `is`("4.00"))
        assertThat(playerSeasonStats[2].gamesPlayed, `is`(10))

    }

    @Test
    fun testFullCareer() {
        val seasonStats = ArrayList<SeasonStats>()
        seasonStats.add(SeasonStats("Winter 16"))
        seasonStats.add(SeasonStats("Fall 17"))

        for (i in 0..4) {
            seasonStats[0].stats.add(GameStats())
            seasonStats[0].stats[i].gameStats = teamStatsForGame

            seasonStats[1].stats.add(GameStats())
            seasonStats[1].stats[i].gameStats = teamStatsForGame
        }

        val player = Player("F")
        player.firstName = "Bob"
        player.lastName = "Builder"
        player.playerID = 2
        player.number = 99

        val playerStats = PlayerStats(2, "bob", "Builder", PlayerPosition.FORWARD)
        playerStats.goals = 2
        playerStats.assists = 4
        playerStats.gamesPlayed = 10
        playerStats.penaltyMinutes = 8

        val careerStatsVM = CareerStatsVM(player, playerStats, seasonStats)

        val playerSeasonStats = careerStatsVM.stats

        assertThat(playerSeasonStats.size, `is`(4))

        assertThat(playerSeasonStats[0].goals, `is`(15))
        assertThat(playerSeasonStats[0].assists, `is`(20))
        assertThat(playerSeasonStats[0].penaltyMinutes, `is`(10))
        assertThat(playerSeasonStats[0].points, `is`(35.toString()))
        assertThat(playerSeasonStats[0].seasonID, `is`("Winter 16"))

        assertThat(playerSeasonStats[1].goals, `is`(15))
        assertThat(playerSeasonStats[1].assists, `is`(20))
        assertThat(playerSeasonStats[1].penaltyMinutes, `is`(10))
        assertThat(playerSeasonStats[1].points, `is`(35.toString()))
        assertThat(playerSeasonStats[1].seasonID, `is`("Fall 17"))

        assertThat(playerSeasonStats[2].goals, `is`(2))
        assertThat(playerSeasonStats[2].assists, `is`(4))
        assertThat(playerSeasonStats[2].penaltyMinutes, `is`(8))
        assertThat(playerSeasonStats[2].points, `is`(6.toString()))
        assertThat(playerSeasonStats[2].seasonID, `is`("Current"))

        assertThat(playerSeasonStats[3].goals, `is`(32))
        assertThat(playerSeasonStats[3].assists, `is`(44))
        assertThat(playerSeasonStats[3].penaltyMinutes, `is`(28))
        assertThat(playerSeasonStats[3].points, `is`(76.toString()))
        assertThat(playerSeasonStats[3].seasonID, `is`("Career"))
    }
}
